aten/src/ATen/native/Col2Im.cpp

```diff
@@ -0,0 +1,334 @@
+      "It is expected dilation equals to 2, but got size ",
+      dilation.size());
+
+  TORCH_CHECK(
+      padding.size() == 2,
+      "It is expected padding equals to 2, but got size ",
+      padding.size());
+
+  TORCH_CHECK(
+      stride.size() == 2,
+      "It is expected stride equals to 2, but got size ",
+      stride.size());
+
+  int64_t output_height = output_size[0];
+  int64_t output_width = output_size[1];
+  int64_t kernel_height = kernel_size[0];
+  int64_t kernel_width = kernel_size[1];
+  int64_t dilation_height = dilation[0];
+  int64_t dilation_width = dilation[1];
+  int64_t pad_height = padding[0];
+  int64_t pad_width = padding[1];
+  int64_t stride_height = stride[0];
+  int64_t stride_width = stride[1];
+
+  col2im_shape_check(
+      input_,
+      Tensor(),
+      output_height,
+      output_width,
+      kernel_height,
+      kernel_width,
+      dilation_height,
+      dilation_width,
+      pad_height,
+      pad_width,
+      stride_height,
+      stride_width);
+
+  Tensor input = input_.contiguous();
+
+  bool batched_input = true;
+  if (input.dim() == 2) {
+    // Force batch
+    batched_input = false;
+    input.resize_({1, input.size(0), input.size(1)});
+  }
+
+  int64_t batch_size = input.size(0);
+  int64_t n_input_plane = input.size(1);
+  int64_t n_output_plane = n_input_plane / (kernel_width * kernel_height);
+
+  output.resize_({batch_size, n_output_plane, output_height, output_width});
+  output.zero_();
+
+  AT_DISPATCH_FLOATING_TYPES_AND_HALF(input.scalar_type(), "col2im_out_cpu", [&] {
+    Tensor input_n = Tensor();
+    Tensor output_n = Tensor();
+
+    int64_t height_col = (output_height + 2 * pad_height -
+                          (dilation_height * (kernel_height - 1) + 1)) /
+            stride_height +
+        1;
+    int64_t width_col = (output_width + 2 * pad_width -
+                         (dilation_width * (kernel_width - 1) + 1)) /
+            stride_width +
+        1;
+
+    for (int64_t elt = 0; elt < batch_size; elt++) {
+      input_n = input.select(0, elt);
+      output_n = output.select(0, elt);
+
+      col2im<scalar_t>(
+          input_n.data<scalar_t>(),
+          n_output_plane,
+          output_height,
+          output_width,
+          height_col,
+          width_col,
+          kernel_height,
+          kernel_width,
+          pad_height,
+          pad_width,
+          stride_height,
+          stride_width,
+          dilation_height,
+          dilation_width,
+          output_n.data<scalar_t>());
+    }
+
+    if (!batched_input) {
+      output.resize_({n_output_plane, output_height, output_width});
+    }
+  });
+}
+
+void col2im_backward_out_cpu_template(
+    Tensor& grad_input,
+    Tensor& grad_output,
+    IntArrayRef kernel_size,
+    IntArrayRef dilation,
+    IntArrayRef padding,
+    IntArrayRef stride) {
+  // im2col_out_cpu checks size of kernel_size, dilation, padding and stride
+  grad_input = im2col_cpu(
```

Comment:
This isn't right. Remember that `Tensor&` input means "write the data into the memory location pointed to by this argument" not "modify this C++ mutable reference". I'm expecting to see you pass in `grad_input` as an input to `im2col_cpu_out`.

Reply:
Ok I didn't create `im2col_cpu_out` because there is no signature for that on `aten/src/ATen/native/native_functions.yaml` should I also create an entry for `out` function there?

Reply:
mmm actually it seems I created the out function just for cuda for col2im .. but actually I am not using that in the code ... I will try to create out functions for col2im and im2col